### PR TITLE
Make 'run line' add a blank line if on last line.

### LINKED ISSUE #3293: F9 does not auto advance when it's at the last line
## Description of your problem

**What steps will reproduce the problem?**

When the cursor is at the last line of code, F9 will execute that line but does not advance to a new (blank) line

**What is the expected output? What do you see instead?**

F9 should always advance to the new blank line given the interactive data analysis workflow. I expect to be able to write new line of code, hit F9 to execute and advance to a new line at the same time. (Rstudio also has this behavior).

Currently, the analyst has to write, hit Enter to advance to a new line, then after a while go back to execute all the lines that have just been written. The third step is unnecessary and breaks the workflow.

## Versions and main components

* Spyder Version: 3.0 beta
* Python Version: 2.7
* Operating system: Ubuntu 14.04

### PR DESCRIPTION
**Pull request by jitseniesen**
Fixes #3293: F9 does not auto advance when it's at the last line. Also add tests.

### CODE REVIEW

File: continuous_integration/appveyor/run_test.bat [hunk 1]
```diff
@@ -26,4 +26,8 @@ echo %time%
 echo Success!
 echo %time%
 echo.
-echo ---------------------------------
+
+echo ------- Running py.test ---------
+echo.
+cd %APPVEYOR_BUILD_FOLDER%
+python runtests.py
```
Comment (ccordoba12) [inline continuous_integration/appveyor/run_test.bat]: Please move this to `appveyor.yml` to have the same configuration as in Travis:

https://github.com/spyder-ide/spyder/blob/master/.travis.yml#L52

**Comment by ccordoba12** [comment]
AppVeyor got stuck while running `spyderlib/utils/tests/test_programs.py` so maybe this test shouldn't be ran on Windows.

**Comment by jitseniesen** [comment]
@ccordabo Thanks. Appveyor should skip those tests now. But it would be good to figure out what is going wrong. Out of curiosity, how can I find out where Appveyor got stuck?

**Comment by ccordoba12** [comment]
@jitseniesen, your last commit revealed an error in our tests. You need to move this import:

https://github.com/spyder-ide/spyder/blob/master/spyderlib/utils/tests/test_syntaxhighlighters.py#L8

after the `qtpy` imports so that the PyQt API is set correctly to API `#2` for PyQt4.

----

Please also check that that's the case for the rest of our tests.

**Comment by jitseniesen** [comment]
@ccordoba12 Thanks, I was wondering what was happening ... I'll have a go at it

**Comment by jitseniesen** [comment]
Phew, all tests past finally. The actual fix for the issue was quite easy, but the tests took me ten times longer! Let's hope it will prove to be a good investment.

@ccordoba12 Moving the `qtpy` import up did not do the trick; it looks like pytest-qt sets PyQt API to v1 on Python 2 before I can do anything about it.

**Comment by ccordoba12** [comment]
Thanks a lot for your hard work on making our tests to work correctly!! I know this kind of job takes a lot of time ;-)

**Comment by goanpeca** [comment]
@jitseniesen awesome :-), it does get easier! and we all have battle scars (some more than others :wink:)

File: runtests.py [hunk 2]
```diff
@@ -8,18 +8,20 @@
 File for running tests programmatically.
 """
 
+# Standard library imports
+import sys
+
 # Third party imports
-import qtpy
 import pytest
```
Comment (ccordoba12) [inline runtests.py]: @jitseniesen, what if we move

    PYTEST_QT_API=pyqt4v2

to here (using `os.environ`, of course)?

That way our tests would run locally too. I mean, right now they only run correctly in Travis and Appveyor :-)

Comment (goanpeca) [inline runtests.py]: And they could run on circle-ci but someone disable that ...

Comment (jitseniesen) [inline runtests.py]: That does not work. But it suggested to me that we should intervene in before the `import pytest` in `runtests.py` so I'm trying that now.

Interesting that you are thinking of running `runtests.py` locally; I find it print too much cruft to be useful.

File: spyderlib/widgets/editor.py [hunk 3]
```diff
@@ -1817,20 +1817,28 @@ def fix_indentation(self, index=None):
...
-            editor.move_cursor_to_next('line', 'down')
+            self.exec_in_extconsole.emit(text, self.focus_to_editor)
+            return
+        editor = self.get_current_editor()
+        line = editor.get_current_line()
+        text = line.lstrip()
+        if text:
+            self.exec_in_extconsole.emit(text, self.focus_to_editor)
+        if editor.is_cursor_on_last_line() and text:
+            editor.append('\n')
```
Comment (ccordoba12) [inline spyderlib/widgets/editor.py]: Please change this to

    editor.append(editor.get_line_separator())

to preserve line endings per operating system.

File: spyderlib/widgets/tests/test_editor.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+def test_run_last_line_when_nonempty(qtbot):
+    editorStack, editor = setup_editor(qtbot)
+    editor.stdkey_backspace() # delete empty line at end
+    old_text = editor.toPlainText()
+    with qtbot.waitSignal(editorStack.exec_in_extconsole) as blocker:
+        editorStack.run_selection()
+    assert blocker.signal_triggered
+    assert blocker.args[0] == 'x = 2'
+    assert editor.toPlainText() == old_text + '\n' # check blank line got added
```
Comment (ccordoba12) [inline spyderlib/widgets/tests/test_editor.py]: This should ne changed too to

    ... == old_text + editor.get_line_separator()

File: spyderlib/widgets/editor.py (hunk 3 above)
Comment (jitseniesen) [inline spyderlib/widgets/editor.py]: Thanks, done

**Pull request merged**